src/pages/dataTableUtilities/reducer/reducerMethods.js

```diff
@@ -0,0 +1,235 @@
+    // Focus first editable cell in next row
+    const nextRowKey = keyExtractor(data[nextRowIndex]);
+    const firstEditableColKey = columns.find(col => col.type === 'editable').key;
+    return setFocus(state, nextRowKey, firstEditableColKey);
+  }
+
+  // We were on the last row and last column, so unfocus current cell
+  return clearFocus(state);
+};
+
+/**
+ * Sets the provided row in action as selected
+ * @param {Object} state  The current state
+ * @param {Object} action The action to act upon
+ * Action: { type: 'selectRow', rowKey }
+ */
+export const selectRow = (state, action) => {
+  const { dataState } = state;
+  const { rowKey } = action;
+  const newDataState = new Map(dataState);
+
+  const rowState = newDataState.get(rowKey);
+  newDataState.set(rowKey, {
+    ...rowState,
+    isSelected: true,
+  });
+
+  return { ...state, dataState: newDataState };
+};
+
+/**
+ * Removes the provided row in action from being
+ * selected.
+ * @param {Object} state  The current state
+ * @param {Object} action The action to act upon
+ * Action: { type: 'deselectRow', rowKey }
+ */
+export const deselectRow = (state, action) => {
+  const { dataState } = state;
+  const { rowKey } = action;
+  const newDataState = new Map(dataState);
+
+  const rowState = newDataState.get(rowKey);
+  newDataState.set(rowKey, {
+    ...rowState,
+    isSelected: false,
+  });
+
+  return { ...state, dataState: newDataState };
+};
+
+/**
+ * Removes all rows from being selected.
+ * @param {Object} state  The current state
+ * Action: { type: 'deselectAll' }
+ */
+export const deselectAll = state => {
+  const { dataState } = state;
+  const newDataState = new Map(dataState);
+
+  // eslint-disable-next-line no-restricted-syntax
+  for (const [rowKey, rowState] of newDataState.entries()) {
+    if (rowState.isSelected) {
+      newDataState.set(rowKey, {
+        ...rowState,
+        isSelected: false,
+      });
+    }
+  }
+  return { ...state, dataState: newDataState };
+};
+/**
+ * Focuses the cell provided in the action passed
+ * @param {Object} state  The current state
+ * @param {Object} action The action to act upon
+ * action: { type: 'focusCell', rowKey, columnKey }
+ */
+export const focusCell = (state, action) => {
+  // Clear any existing focus and focus cell specified in action
+  const { rowKey, columnKey } = action;
+  return setFocus(state, rowKey, columnKey);
+};
+
+/**
+ * Sorts the current set of data by the provided
+ * key and direction in action.
+ * @param {Object} state  The current state
+ * @param {Object} action The action to act upon
+ * Action: { type: 'sortData', sortBy }
+ * sortBy: String key of the field of the objects to sort by.
+ */
+export const sortData = (state, action) => {
+  const { data, isAscending } = state;
+  const { sortBy } = action;
+  const columnKeyToDataType = {
+    itemCode: 'string',
+    itemName: 'string',
+    availableQuantity: 'number',
+    totalQuantity: 'number',
+  };
```

Comment:
can come later, but this needs to be extensible?

Reply:
100% - I am thinking at this point, just a constant in this file, any other suggestions? Potentially larger reducer methods could have their own file.. but that's pretty ugly

Reply:
Later it is, saving my thought power